Revise the "Logical AND (&&)" page (#10846)
mdn/content: files/en-us/web/javascript/reference/operators/logical_and/index.md

---
{{jsSidebar("Operators")}}

> The logical AND (`&&`) operator (logical conjunction) for a set of operands is true if, and only if, all of its operands are true.

Reviewer: I quite like this option (though you will need to clean up the following bits)

```suggestion
The logical AND (`&&`) operator (logical conjunction) for a set of {{jsxref("Boolean")}} operands will be `true` if and only if all the operands are `true`. Otherwise it will be `false`.

More generally, the operator returns the value of the first {{Glossary("falsy")}} operand encountered when evaluating from left to right, or the value of the last operand if they are all {{Glossary("truthy")}}.
```

Reviewer: And if you took that suggestion you would remove the two lines below

{{jsSidebar("Operators")}}

The logical AND (`&&`) operator (logical conjunction) for a set of
operands is true if, and only if, all of its operands are true. It is typically used with
{{jsxref("Boolean")}} (logical) values, in which case it returns a Boolean value. However,

> when the `&&` operator is used with non-Boolean values, it preserves the value of one of the 

Reviewer: This is not "quite" right/complete because `is used with non-Boolean values` implies all non-Boolean values, while the more general case is a mixture of boolean and non-boolean values. 

I'll see if I can tidy this.

Author: I slightly misinterpreted the original definition, but if I've understood correctly now:

> The logical AND (`&&`) operator (logical conjunction) for a set of operands is true if, and only if, all of its operands are true. This operator returns the value of one of its specified operands and is typically used with {{jsxref("Boolean")}} (logical) values, in which case it returns a Boolean value. If the `&&` operator is used with non-Boolean values, it preserves the value of one of the operands and returns it as the original non-Boolean value.

It may need some syntactical improvements and a clearer explanation, but would something like the above be correct/complete?

Reviewer: The operator _always_ returns the value of an operand. It's just that in the non-boolean case this operand may be non-boolean. The definition above falls down a bit if you mix comparison of booleans and non booleans.

The other aspect is that we talk here about converting and preserving but I doubt that the value is actually converted or preserved. It is more like each operand is evaluated in turn to get a value. If that value evaluates as falsy the expression returns the value. Otherwise it evaluates the next operand. Hope that the distinction is obvious.

I am being a bit nitpicky. This is in my opinion already better than what was there. 

I'll leave you to think about this a bit more. I had a play and it is not so easy. Note I will be back Friday and can comment a bit more then. @wbamberg is likely to have comments too.

Author: I see. I'll do my best to correct and improve this. Thank you for clarifying!

## Description

> Logical AND (`&&`) evaluates operands from left to right, converts each to a Boolean, then returns the first falsy value found; if all values are truthy, the last value is returned.

Reviewer: Perhaps 
```suggestion
Logical AND (`&&`) evaluates operands from left to right, returning the value of the first {{Glossary("falsy")}} operand; if all values are {{Glossary("truthy")}}, the value of the last operand is returned.
```

If a value can be converted to `true`, the value is so-called {{Glossary("truthy")}}. If a value can be converted to `false`, the value is so-called {{Glossary("falsy")}}.

Examples of expressions that can be converted to false are:

- `false`;
- `null`;
- `NaN`;
- `0`;
- empty string (`""` or `''` or ` `` `);
- `undefined`.

> The AND operator preserves non-Boolean values and returns them as is:

Reviewer: Do you mean preserves Boolean values - i.e. do you really mean non-Boolean?

Author: I believe I meant to say that the `&&` will return non-Boolean values as they are (i.e. falsy value `0` is returned as `0` and not `false`, if I'm not mistaken). And it seems I may have caught a grammatical error there, though research has given me mixed results—I'm not quite sure whether it should be "as they are" or "as is".

This probably goes back to the definition, where we were discussing the return value of the `&&` operator, so perhaps this could be changed once a complete definition for this operator has been accepted? Or should this be changed directly to "...preserves Boolean values"?

Reviewer: I'm wrong! I looked at the first example - that is comparison of 0 and 1 which I read as booleans, because they map directly to booleans in languages I am more familiar with. So this is correct.

BUT you might want to consider not using numbers for at least the first case

```js

> result = 0 && 1;   // result is assigned 0

Author: ```suggestion
result = '' && 'foo';   // result is assigned '' (empty string)
```

does **not** evaluate any of the remaining operands:

`(some falsy expression) && expr` is short-circuit
evaluated to the falsy expression;

Due to short-circuiting, the `expr` part above is **never evaluated**, 
meaning any side effects of doing so won't occur (e.g., if `expr` is a 
function call, the calling never takes place). This happens because the
first operand evaluated is converted to `false`, which turns the final value 
of the AND operator into `(some falsy expression)` if it's non-Boolean, or 
`false` if it's a Boolean. See the example below:

```js
function A() { console.log('called A'); return false; }
function B() { console.log('called B'); return true; }

console.log( A() && B() );
// logs "called A" to the console due to the call for function A,
// && evaluates to false (function A returns false), then false is logged to the console;
// the AND operator short-circuits here and ignores function B
```

### Operator precedence

The AND operator has a higher precedence than the OR operator, meaning the `&&` 
operator is executed before the `||` operator (see [operator precedence](/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence)).

```js

> false || true && true             // evaluates `true && true` first, then the || finds and returns true

Reviewer: The comment doesn't seem quite right "the || finds and returns true"

Author: Wasn't quite sure how to word this, and I noticed that I wrote this in a comment below (but with "finds and returns false") as well. If the third comment "evaluates `((4 < 0) && (1 == 1))` first, becomes (2 == 3) || false, then returns false" is acceptable, would changing the other two to a similar format work?

For example, replacing "the || finds and returns true" with something like "becomes false || true, then returns true"?

Reviewer: This is complicated because we're trying to explain what happens first. I think there are two easy options:

- We just state the result - they can work out "how" it works themselves.

   ```js
   false || true && true             // Returns true
   ```
- We _show_ how they evaluate:

   ```js
   false || true && true             // Returns true. This is the same as "false || true".
   (2 == 3) || ((4 < 0) && (1 == 1)) // returns false. This is the same as false || (false && true) = false || false.
   ```

Author: I suppose the first option would be enough, since there is a separate page on operator precedence? 

And on line 95, would it be alright to remove the parentheses around `(4 < 0) && (1 == 1)`? They aren't necessary for operator precedence to apply but I made a mistake when I was correcting an extra `)` there. (Apologies for all these questions and comments—this is the first time I'm writing or revising documentation)

Reviewer: I'm going to leave that to you. It is a judgement call and neither will be wrong. 

There is no need to apologize. For me this is an investment :-)
